network/network-address/src/lib.rs

```diff
@@ -0,0 +1,619 @@
+
+const MAX_DNS_NAME_SIZE: usize = 255;
+
+/// A `RawNetworkAddress` is the serialized, unverified, on-chain representation
+/// of a [`NetworkAddress`]. Specifically, a `RawNetworkAddress` is usually an
+/// [`lcs`]-serialized `NetworkAddress`.
+///
+/// This representation is useful because:
+///
+/// 1. Move does't understand (and doesn't really need to understand) what a
+///    `NetworkAddress` is, so we can just store an opaque `vector<u8>` on-chain,
+///    which we represent as `RawNetworkAddress` in Rust.
+/// 2. We want to deserialize a `Vec<NetworkAddress>` but ignore ones that don't
+///    properly deserialize. For example, a validator might advertise several
+///    `NetworkAddress`. If that validator is running a newer version of the node
+///    software, and old versions can't understand one of the `NetworkAddress`,
+///    they would be able to ignore the one that doesn't deserialize for them.
+///    We can easily do this by storing a `vector<vector<u8>>` on-chain, which
+///    we deserialize as `Vec<RawNetworkAddress>` in Rust. Then we deserialize
+///    each `RawNetworkAddress` into a `NetworkAddress` individually.
+///
+/// Note: deserializing a `RawNetworkAddress` does no validation, other than
+/// deserializing the underlying `Vec<u8>`.
+#[derive(Clone, Debug, Eq, PartialEq, Deserialize, Serialize)]
+pub struct RawNetworkAddress(#[serde(with = "serde_bytes")] Vec<u8>);
+
+/// Libra `NetworkAddress` is a compact, efficient, self-describing and
+/// future-proof network address represented as a stack of protocols. Essentially
+/// libp2p's [multiaddr](https://multiformats.io/multiaddr/) but using [`lcs`] to
+/// describe the binary format.
+///
+/// Most validators will advertise a network address like:
+///
+/// `/dns/example.com/tcp/6180/ln-noise-ik/<x25519-pubkey>/ln-handshake/1`
+///
+/// Unpacking, the above effectively means:
+///
+/// 1. Resolve the DNS name "example.com" to an ip address, `addr`.
+/// 2. Open a TCP connection to `(addr, 6180)`.
+/// 3. Perform a Noise IK handshake and assume the peer's static pubkey is
+///    `<x25519-pubkey>`. After this step, we will have a secure, authenticated
+///    connection with the peer.
+/// 4. Perform a LibraNet version negotiation handshake (version 1).
+///
+/// One key concept behind `NetworkAddress` is that it is fully self-describing,
+/// which allows us to easily "pre-negotiate" protocols while also allowing for
+/// future upgrades. For example, it is generally unsafe to negotiate a secure
+/// transport in-band. Instead, with `NetworkAddress` we can advertise (via
+/// discovery) the specific secure transport protocol and public key that we
+/// support (and even advertise multiple incompatible versions). When a peer
+/// wishes to establish a connection with us, they already know which secure
+/// transport protocol to use; in this sense, the secure transport protocol is
+/// "pre-negotiated" by the dialier selecting which advertised protocol to use.
```

Comment:
I don't think this mitigates downgrade attacks. For example if I advertise two different protocols at two different addresses, and A tries to connect to me but can't with one protocol (maybe a MITM is blocking the connection) they might attempt to try to use the second protocol on the other address.